### DISCUSSION TITLE
Proper method to update templated "git-version.h.in" file

### DISCUSSION THREAD

**Discussion by tabedzki** [mesonbuild/meson#10896] [Q&A] [answered]
Hi there, 
I am trying to get the git hash of a project written to a header file for c++. I have a template `src/git-version.h.in` and a intended target `src/git-version.h`. 
I've read the documentation for `vcs_tag` and I've looked at a few StackOverflows but I am unable to get it working. The current error I get is 
`src/meson.build:56:0: ERROR: vcs_tag keyword argument "output" Output 'src/git_version.h' must not contain a path segment.`

Thanks in advance for your help/suggestions.

```
./meson.build
./src/meson.build
./src/main.cpp (this has the command to include the header file)
./src/git-version.h
```

meson.build
```
project('cuda-example', ['cuda', 'cpp'],
    version         : '0.1.0',
    license         : 'MIT',
    meson_version   : '>=0.58.0',
    default_options : ['warning_level=2','cuda_std=c++11','c_std=c11', 'cpp_std=c++11'])

nvcc = meson.get_compiler('cuda')

cuda = import('unstable-cuda')

subdir('src')
...
executable(
  target_exec,
  gpu_files,
 dependencies : [dep, vcs_dep], install : true, cuda_args: [myargs] 
 )
```

src/meson.build
```
gpu_files=['src/main.cpp']


vcs_dep=vcs_tag(command: ['git', 'describe', '--tags', '--always', '--dirty', '--abbrev=9'], 
      input:'src/git_version.h.in',
      output:'src/git_version.h',
      replace_string:'@GIT_VERSION@')
```

src/main.cu
```
#include "git-version.h"
int main(int argc, char** argv)
{
	printf("Git Version hash: %s\n", MY_GIT_VERSION);
        return 0;
}
```

src/git-version.h.in
```
#define MY_GIT_VERSION "@GIT_VERSION@"
```

**Comment by eli-schwartz** [comment, marked as answer]
Specify the output file as
```meson
output: 'git_version.h'
```

Meson will automatically treat that as relative to the current build directory context, which is `src/` because it is defined in `src/meson.build`.

You should also do this for the input file as well, though, because otherwise it won't find the input file.

The source file defined as a list of strings can be defined instead with `files('main.cpp')` and it will remember the directory it was defined in, but if you leave it as a plain string then it's "defined" at the time of *use*, which is in the root meson.build

**Reply by tabedzki** [reply]
thanks very much!
Also, the `./meson.build` should be 

```
executable(
  target_exec,
  gpu_files,
 dependencies : dep, install : true, cuda_args: [myargs] 
 )
 ```
 
 and `./src/meson.build` should be 
 
 
 ```
 vcs_dep = vcs_tag(command: ['git', 'describe', '--tags', '--always', '--dirty', '--abbrev=9'], 
      input: 'git_version.h.in',
      output:'git_version.h',
      replace_string:'@GIT_VERSION@')

 gpu_files=files(['main.cpp'])
 gpu_files += vcs_dep
 ```

**Reply by tabedzki** [reply]
@eli-schwartz  I have a follow up question. How do I get `meson` to (1)  check if the `vcs_tag` is old, and if it is old (2) update `git_version.h`  and (2.1) force recompilation of main.cpp?
